### ISSUE TITLE
Start service mssql-server error in Ubuntu 18.04 (WSL2) - System has not been booted with systemd as init system

### ISSUE THREAD

**Issue by marcelolaias** [MicrosoftDocs/WSL#815]
I followed the tutorial on the link [https://docs.microsoft.com/en-us/sql/linux/quickstart-install-connect-ubuntu?view=sql-server-ver15](url) to install SQL Server 2019 on Ubuntu 18.04 on WSL2.

After running the command `sudo/opt/mssql/bin/mssql-conf setup` the following error occurs:

usermod: no changes
Choose an edition of SQL Server:
  1) Evaluation (free, no production use rights, 180-day limit)
  2) Developer (free, no production use rights)
  3) Express (free)
  4) Web (PAID)
  5) Standard (PAID)
  6) Enterprise (PAID) - CPU Core utilization restricted to 20 physical/40 hyperthreaded
  7) Enterprise Core (PAID) - CPU Core utilization up to Operating System Maximum
  8) I bought a license through a retail sales channel and have a product key to enter.

Details about editions can be found at
https://go.microsoft.com/fwlink/?LinkId=2109348&clcid=0x409

Use of PAID editions of this software requires separate licensing through a
Microsoft Volume Licensing program.
By choosing a PAID edition, you are verifying that you have the appropriate
number of licenses in place to install and run this software.

Enter your edition(1-8): 2
The license terms for this product can be found in
/usr/share/doc/mssql-server or downloaded from:
https://go.microsoft.com/fwlink/?LinkId=2104294&clcid=0x409

The privacy statement can be viewed at:
https://go.microsoft.com/fwlink/?LinkId=853010&clcid=0x409

Do you accept the license terms? [Yes/No]:Yes

Enter the SQL Server system administrator password:
Confirm the SQL Server system administrator password:
Configuring SQL Server...

ForceFlush is enabled for this instance.
ForceFlush feature is enabled for log durability.
DBSTARTUP (msdb, 4): FCBOpenTime took 437 ms
DBSTARTUP (msdb, 4): FCBHeaderReadTime took 368 ms
DBSTARTUP (msdb, 4): FileMgrPreRecoveryTime took 977 ms
DBSTARTUP (msdb, 4): MasterFilesScanTime took 582 ms
DBSTARTUP (msdb, 4): UpgradeTime took 13542 ms
DBSTARTUP (msdb, 4): StartupInDatabaseTime took 13655 ms
DBSTARTUP (msdb, 4): RemapSysfiles1Time took 103 ms
System has not been booted with systemd as init system (PID 1). Can't operate.
Attempting to start the Microsoft SQL Server service failed.
---

I tried to run the `sudo reboot` command and a similar error occurred:

neologics@w10delli7k:~$ sudo reboot
[sudo] password for neologics:
System has not been booted with systemd as init system (PID 1). Can't operate.
Failed to talk to init daemon.

Note that this repo's issue tracker is only for reporting issues with the WSL docs.

Other related issue repo's:

* WSL: https://github.com/Microsoft/BashOnWindows/
* Windows Console: https://github.com/Microsoft/Console/

**Comment by zodraz** [comment]
Same issue here...

**Comment by tleung927** [comment]
same here. google for an hour and wsl didn't have systemctl need to use service xxxx start instead. But got other error show mssql-server unrecognized service .......

**Comment by averyrobbins1** [comment]
I got this same error. Has anyone figured it out yet?

**Comment by brucedkyle** [comment]
Looking for a work around too

**Comment by aymacdonald** [comment]
I haven't done anything with it yet, but I have gotten MSSQL running on WSL2.  As you've noticed, the installer package only supports configuring systemd and therefore "can't operate".

this command "service start mssql-server" is for System V service init scripts, so the unrecognized service would be expected.  In order for that to work, you first would need to create a file named mssql-server, with executable permissions, in /etc/init.d

This will start it.  
`sudo -u mssql /opt/mssql/bin/sqlservr -c \`
`    -d/var/opt/mssql/data/master.mdf \`
`    -l/var/opt/mssql/data/mastlog.ldf \`
`    -e/var/opt/mssql/log/errorlog \`
`    -x`

Info on the command-line parameters of [sqlservr](https://docs.microsoft.com/en-us/sql/tools/sqlservr-application?view=sql-server-linux-ver15)

**Comment by feslima** [comment]
@averyrobbins1 @brucedkyle  I was able to make it work following [this post](https://blog.developpez.com/mikedavem/p13196/sql-server-2012/dealing-with-sql-server-on-linux-on-wsl2).

**TL;DR:**
1. Create a file named _mssql-server_ in the folder _/etc/init.d_:
`sudo nano /etc/init.d/mssql-server`
2. Paste this code into it:
```
#! /bin/sh -e
#
### BEGIN INIT INFO
# Provides:          sqlserver
# Required-Start:    $all
# Required-Stop:
# Default-Start:     2 3 4 5
# Default-Stop:      0 1 6
# Short-Description: Manages SQL Server instance on Linux
### END INIT INFO

DAEMON="/opt/mssql/bin/sqlservr"
daemon_OPT=""
DAEMONUSER="mssql"
daemon_NAME="sqlservr"

export PATH=/sbin:/bin:/usr/sbin:/usr/bin:/usr/local/sbin:/usr/local/bin:/var/opt/mssql

# Check sqlserver is present
if [ ! -x $DAEMON ]; then
        log_failure_msg "$DAEMON not present or not executable"
        exit 1
fi

# Load init functions
. /lib/lsb/init-functions


d_start () {
        log_daemon_msg "Starting system $daemon_NAME Daemon"
        start-stop-daemon --background --name $daemon_NAME --start --quiet --chuid $DAEMONUSER --exec $DAEMON --umask 007 -- $DAEMON_OPTS --oknodo
        log_end_msg $?
}

d_stop () {
        log_daemon_msg "Stopping system $daemon_NAME Daemon"
        start-stop-daemon --name $daemon_NAME --stop --retry 5 --quiet --name $daemon_NAME --oknodo
        log_end_msg $?
}

case "$1" in

        start|stop)
                d_${1}
                ;;

        restart|reload|force-reload)
                        d_stop
                        d_start
                ;;

        force-stop)
               d_stop
                ;;

        status)
                status_of_proc "$daemon_NAME" "$DAEMON" "system-wide $daemon_NAME" && exit 0 || exit $?
                ;;
        *)
                echo "Usage: /etc/init.d/$daemon_NAME {start|stop|force-stop|restart|reload|force-reload|status}"
                exit 1
                ;;
esac
exit 0
```
_Note that at line 57 in the link above there is a formatting error (it was `$daemon_NAME" &amp;&amp; exit`, changed to `$daemon_NAME" && exit`)._

3. Edit the file permissions with `sudo chmod 755 /etc/init.d/mssql-server`;
4. Now you are able to check the service status with: `sudo service mssql-server status`;
5. You have to start/stop the service manually, though. Commands are:
`sudo service mssql-server start` and `sudo service mssql-server stop`

**Comment by vivekgoyal923** [comment]
Thanks @feslima and @aymacdonald.

For me, somehow, running the command 
`sudo service mssql-server start` starts and stops the sql server.

But, following command successfully starts the server after adding mssql-server file /etc/init.d
`sudo -u mssql /opt/mssql/bin/sqlservr -c \
-d/var/opt/mssql/data/master.mdf \
-l/var/opt/mssql/data/mastlog.ldf \
-e/var/opt/mssql/log/errorlog \
-x`

Just wanted to mention if someone faces similar issue in future.

**Comment by JAlfonsoRG** [comment]
Hello. The solution that provides @feslima works for me, thank you. The only thing that I modified was to remove the `-- $DAEMON_OPTS` option from line 31, because that variable it's not declared and causes that the service stops suddenly.

**Comment by Sotiris-k** [comment]
I stumbled upon this problem myself and @feslima 's solution works but you have to try and run the command to start the service yourself first because the problems aren't logged anywhere by default and it makes you believe that it's the script's fault while it isn't.

**Comment by codigoperfeito** [comment]
> @averyrobbins1 @brucedkyle I was able to make it work following [this post](https://blog.developpez.com/mikedavem/p13196/sql-server-2012/dealing-with-sql-server-on-linux-on-wsl2).
> 
> **TL;DR:**
> 
> 1. Create a file named _mssql-server_ in the folder _/etc/init.d_:
>    `sudo nano /etc/init.d/mssql-server`
> 2. Paste this code into it:
> 
> ```
> #! /bin/sh -e
> #
> ### BEGIN INIT INFO
> # Provides:          sqlserver
> # Required-Start:    $all
> # Required-Stop:
> # Default-Start:     2 3 4 5
> # Default-Stop:      0 1 6
> # Short-Description: Manages SQL Server instance on Linux
> ### END INIT INFO
> 
> DAEMON="/opt/mssql/bin/sqlservr"
> daemon_OPT=""
> DAEMONUSER="mssql"
> daemon_NAME="sqlservr"
> 
> export PATH=/sbin:/bin:/usr/sbin:/usr/bin:/usr/local/sbin:/usr/local/bin:/var/opt/mssql
> 
> # Check sqlserver is present
> if [ ! -x $DAEMON ]; then
>         log_failure_msg "$DAEMON not present or not executable"
>         exit 1
> fi
> 
> # Load init functions
> . /lib/lsb/init-functions
> 
> 
> d_start () {
>         log_daemon_msg "Starting system $daemon_NAME Daemon"
>         start-stop-daemon --background --name $daemon_NAME --start --quiet --chuid $DAEMONUSER --exec $DAEMON --umask 007 -- $DAEMON_OPTS --oknodo
>         log_end_msg $?
> }
> 
> d_stop () {
>         log_daemon_msg "Stopping system $daemon_NAME Daemon"
>         start-stop-daemon --name $daemon_NAME --stop --retry 5 --quiet --name $daemon_NAME --oknodo
>         log_end_msg $?
> }
> 
> case "$1" in
> 
>         start|stop)
>                 d_${1}
>                 ;;
> 
>         restart|reload|force-reload)
>                         d_stop
>                         d_start
>                 ;;
> 
>         force-stop)
>                d_stop
>                 ;;
> 
>         status)
>                 status_of_proc "$daemon_NAME" "$DAEMON" "system-wide $daemon_NAME" && exit 0 || exit $?
>                 ;;
>         *)
>                 echo "Usage: /etc/init.d/$daemon_NAME {start|stop|force-stop|restart|reload|force-reload|status}"
>                 exit 1
>                 ;;
> esac
> exit 0
> ```
> 
> _Note that at line 57 in the link above there is a formatting error (it was `$daemon_NAME" &amp;&amp; exit`, changed to `$daemon_NAME" && exit`)._
> 
> 3. Edit the file permissions with `sudo chmod 755 /etc/init.d/mssql-server`;
> 4. Now you are able to check the service status with: `sudo service mssql-server status`;
> 5. You have to start/stop the service manually, though. Commands are:
>    `sudo service mssql-server start` and `sudo service mssql-server stop`

thank u so much, here is perfect !

**Comment by mattwojo** [comment]
Now that [SystemD is supported](https://devblogs.microsoft.com/commandline/systemd-support-is-now-available-in-wsl/) on WSL 2, I believe this issue should be resolved, so I'm going to close it and update the doc... however, please re-open if you find that this is not the case.
